cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/DefaultPluginConfigurer.java

```diff
@@ -149,7 +159,58 @@ private Plugin findPlugin(String pluginType, String pluginName, String pluginId,
     Preconditions.checkArgument(!plugins.containsKey(pluginId),
                                 "Plugin of type %s, name %s was already added as id %s.",
                                 pluginType, pluginName, pluginId);
-    return FindPluginHelper.findPlugin(artifactRepository, pluginInstantiator, deployNamespace.toEntityId(), artifactId,
-                                       pluginType, pluginName, properties, selector);
+
+
+    final Class[] callerClasses = CallerClassSecurityManager.getCallerClasses();
+    if (callerClasses.length < 3) {
+      // This shouldn't happen as there should be someone calling this method.
+      throw new IllegalStateException("Invalid call stack.");
+    }
+
+    LinkedHashSet<ArtifactId> parents = new LinkedHashSet<>();
+    // go through the call stack to get the parent plugins. This is to support plugins of plugins.
+    // for example, suppose appX uses pluginY, which uses pluginZ. When we try to use pluginZ, we need to first
+    // check if pluginZ extends pluginY. If it does not, we then check if it extends appX.
+    // 0 is the CallerClassSecurityManager, 1 is this class, hence 2 is the actual caller
+    for (int i = 2; i < callerClasses.length; i++) {
+      ClassLoader callerClassloader = callerClasses[i].getClassLoader();
+      if (AbstractPluginConfigurable.class.isAssignableFrom(callerClasses[i])) {
```

Comment:
I think it's better not to terminate early. E.g. one can have an application that uses plugin to load another application, which in turn load another application via plugin